library/init/category/lift.lean

```diff
@@ -0,0 +1,78 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gabriel Ebner, Sebastian Ullrich
+-/
+prelude
+import init.function init.coe
+import init.category.monad
+import init.util
+
+universes u v w
+
+/-- A function for lifting a computation from an inner monad to an outer monad.
+    Like https://hackage.haskell.org/package/transformers-0.5.5.0/docs/Control-Monad-Trans-Class.html,
+    but `n` does not have to be a monad transformer. -/
+class has_monad_lift (m : Type u → Type v) (n : Type u → Type w) :=
+(monad_lift {} : ∀ {α}, m α → n α)
+
+/-- The reflexive-transitive closure of `has_monad_lift`. -/
+class has_monad_lift_t (m : Type u → Type v) (n : Type u → Type w) :=
+(monad_lift {} : ∀ {α}, m α → n α)
+
+export has_monad_lift_t (monad_lift)
+
+@[reducible] def has_monad_lift_to_has_coe {m n} [has_monad_lift_t m n] {α} : has_coe (m α) (n α) :=
+⟨monad_lift⟩
+
+instance has_monad_lift_t_trans (m n o) [has_monad_lift n o] [has_monad_lift_t m n] : has_monad_lift_t m o :=
+⟨λ α ma, has_monad_lift.monad_lift (monad_lift ma : n α)⟩
+
+instance has_monad_lift_t_refl (m) : has_monad_lift_t m m :=
+⟨λ α, id⟩
+
+@[simp] lemma monad_lift_refl {m : Type u → Type v} {α} : (monad_lift : m α → m α) = id := rfl
+
+
+/-- A functor in the category of monads. Can be used to lift monad-transforming functions.
+    Based on https://hackage.haskell.org/package/pipes-2.4.0/docs/Control-MFunctor.html,
+    but not restricted to monad transformers. -/
+class monad_functor (m m' : Type u → Type v) (n n' : Type u → Type w) :=
+(monad_map {} {α : Type u} : (∀ {α}, m α → m' α) → n α → n' α)
+
+/-- The reflexive-transitive closure of `monad_functor`. -/
+class monad_functor_t (m m' : Type u → Type v) (n n' : Type u → Type w) :=
+(monad_map {} {α : Type u} : (∀ {α}, m α → m' α) → n α → n' α)
+
+export monad_functor_t (monad_map)
+
+instance monad_functor_t_trans (m m' n n' o o') [monad_functor n n' o o'] [monad_functor_t m m' n n'] :
+  monad_functor_t m m' o o' :=
+⟨λ α f, monad_functor.monad_map (λ α, (monad_map @f : n α → n' α))⟩
+
+instance monad_functor_t_refl (m m') : monad_functor_t m m' m m' :=
+⟨λ α f, f⟩
+
+@[simp] lemma monad_map_refl {m m' : Type u → Type v} (f : ∀ {α}, m α → m' α) {α} : (monad_map @f : m α → m' α) = f := rfl
+
+
+/-- Run a monad stack to completion. -/
+class monad_run (out : out_param $ Type u → Type v) (m : Type u → Type v) :=
+(run {} {α : Type u} : m α → out α)
+(unrun {} {α : Type u} : out α → m α)
+
+export monad_run (run unrun)
+
+
+/-- Lift an impure scoping function.
+    The helper functions in init.util take a thunk of a pure computation and execute it under some side
+    effect. For monads like `state_t` that are "lazy", i.e. evaluate to a closure, we have to special
+    case these operations to go inside these binders and work on the strict "core". -/
+class has_scope_impure (m : Type u → Type v) :=
+(scope_impure_opt {} {α : Type u} : (∀ {β : Type u}, thunk β → option β) → thunk (m α) → m (option α))
```

Comment:
It is not clear to me what `scope_impure_opt f t` is supposed to do when I read this definition.
I had to look at actual instances to try to understand it.
I also had to look at `init/util.lean` to (try to) understand why you use the `option` type.

Reply:
BTW, does `{β : Type u}` buy us anything? I think it just forces us to use `@f` when implementing the instances. AFAICT, the only place where it was useful is at the definition of `has_scope_impure id`, where you could write `f (x ())` instead of `f _ (x ())`.

Reply:
This type class is still quite obscure to me. I am inclined to sacrifice `init.util` to avoid this extra complexity.
We can provide the functions defined at `init.util` as actions in the `tactic` and `io` monads.